qrenderdoc/Windows/Dialogs/PerformanceCounterSelection.cpp

```diff
@@ -0,0 +1,290 @@
+/******************************************************************************
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016-2017 Baldur Karlsson
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "PerformanceCounterSelection.h"
+#include "Code/CaptureContext.h"
+#include "Code/Interface/QRDInterface.h"
+#include "ui_PerformanceCounterSelection.h"
+
+#include <set>
+#include <unordered_map>
+
+#define JSON_ID "rdocPerformanceCounterSettings"
+#define JSON_VER 1
+
+namespace
+{
+enum class CounterFamily
+{
+  Unknown,
+  Generic,
+  AMD,
+  Intel,
+  NVIDIA,
+};
+
+CounterFamily GetCounterFamily(GPUCounter counter)
+{
+  if(IsAMDCounter(counter))
+  {
+    return CounterFamily::AMD;
+  }
+  else if(IsIntelCounter(counter))
+  {
+    return CounterFamily::Intel;
+  }
+  else if(IsNvidiaCounter(counter))
+  {
+    return CounterFamily::NVIDIA;
+  }
+
+  return CounterFamily::Generic;
+}
+
+const char *ToString(CounterFamily family)
+{
+  switch(family)
+  {
+    case CounterFamily::AMD: return "AMD";
+    case CounterFamily::Generic: return "Generic";
+    case CounterFamily::Intel: return "Intel";
+    case CounterFamily::NVIDIA: return "NVIDIA";
+    case CounterFamily::Unknown: return "Unknown";
+  }
+
+  return nullptr;
+}
+}
+
+PerformanceCounterSelection::PerformanceCounterSelection(ICaptureContext &ctx, QWidget *parent)
+    : QDialog(parent), ui(new Ui::PerformanceCounterSelection), m_Ctx(ctx)
+{
+  ui->setupUi(this);
+
+  connect(ui->counterTree, &QTreeWidget::itemEntered, [&](QTreeWidgetItem *item, int) -> void {
+    const auto d = item->data(0, Qt::UserRole + 1);
+
+    if(d.isValid())
+    {
+      ui->counterDescription->setText(
+          QString(QLatin1String("<b>%1</b><hr>%2")).arg(item->text(0)).arg(d.toString()));
+    }
+  });
+
+  connect(ui->save, &QPushButton::pressed, [&]() -> void {
```

Comment:
I'd rather use the normal signal/slot mechanism for cases like this where it's a simple handler for a button created in the UI itself with nothing that changes at runtime. Otherwise you end up with a constructor with all of the code in it in various lambdas and connects, and no real organisation.